syntax = "proto3";

package simulate;

// We move all of the hard work of actually representing an arbitrary gearset
// as an equipment summary into the scripting side, rather than the engine.
// This allows us to pass the engine a very compact representation of equipment
// which provides a partial ordering.
message EquipmentSummary {
  int32 strength = 1;
  int32 intelligence = 2;
  int32 agility = 3;
  int32 stamina = 4;
  int32 mastery_rating = 5;
  int32 versatility_rating = 6;
  int32 haste_rating = 7;
  int32 crit_rating = 8;
  int32 attack_power = 9;
  int32 spell_power = 10;
  double mainhand_weapon_speed = 11;
  double mainhand_weapon_damage_dps = 12;
  double offhand_weapon_speed = 13;
  double offhand_weapon_damage_dps = 14;

  // Things that modify sim behavior, like trinket / enchant procs, or
  // legendary static effects, or "replacement effects" of any kind.
  repeated int32 equip_effects = 15;

  // Things that provide additional "action space" for the player.
  repeated int32 on_use_effects = 16;
}

message TalentConfig {
  int32 row_1 = 1;
  int32 row_2 = 2;
  int32 row_3 = 3;
  int32 row_4 = 4;
  int32 row_5 = 5;
  int32 row_6 = 6;
  int32 row_7 = 7;
}

enum Specialization {
  UNUSED_SPECIALIZATION = 0;
  DRUID_FERAL = 1;
  DRUID_BALANCE = 2;
  MAGE_FIRE = 3;
  MAGE_FROST = 4;
  MAGE_ARCANE = 5;
  WARLOCK_DEMONOLOGY = 6;
  WARLOCK_DESTRUCTION = 7;
  WARLOCK_AFFLICTION = 8;
  WARRIOR_ARMS = 9;
  WARRIOR_FURY = 10;
  DEATH_KNIGHT_FROST = 11;
  DEATH_KNIGHT_UNHOLY = 12;
  PALADIN_RETRIBUTION = 13;
  HUNTER_BEAST_MASTERY = 14;
  HUNTER_SURVIVAL = 15;
  HUNTER_MARKSMANSHIP = 16;
  PRIEST_SHADOW = 17;
  SHAMAN_ELEMENTAL = 18;
  SHAMAN_ENHANCEMENT = 19;
  MONK_WINDWALKER = 20;
  DEMON_HUNTER_HAVOC = 23;
  ROGUE_ASSASSINATION = 24;
  ROGUE_OUTLAW = 25;
  ROGUE_SUBTLETLY = 26;
}

enum Race {
  UNUSED_RACE = 0;
  HUMAN = 1;
  DWARF = 2;
  NIGHT_ELF = 3;
  GNOME = 4;
  ORC = 5;
  TROLL = 6;
  UNDEAD = 7;
  TAUREN = 8;
  DRAENEI = 9;
  BLOOD_ELF = 10;
  WORGEN = 11;
  GOBLIN = 12;
  PANDAREN = 13;
}

message Consumables {
  int32 flask_id = 1;
  int32 food_id = 2;
  int32 augment_id = 3;
  int32 prepot_id = 4;
}

message PlayerConfig {
  string name = 1;
  Race race = 2;
  Specialization spec = 3;
  TalentConfig talents = 4;
  EquipmentSummary equipment_summary = 5;
  Consumables consumables = 6;
}
